docs/library/machine.ADCBlock.rst

```diff
@@ -0,0 +1,57 @@
+.. currentmodule:: machine
+.. _machine.ADCBlock:
+
+class ADCBlock -- control ADC peripherals
+=========================================
+
+The ADCBlock class provides access to an ADC peripheral which has a
+number of channels that can be used to sample analog values.  It allows
+finer control over configuration of :ref:`machine.ADC <machine.ADC>`
+objects, which do the actual sampling.
+
+Example usage::
+
+   import machine
+
+   block = machine.ADCBlock(id, 12) # create an ADCBlock with 12-bit resolution
+   adc = block.connect(4, pin)      # connect channel 4 to the given pin
+   val = adc.read_v()               # read an analog value
+
+Constructors
+------------
+
+.. class:: ADCBlock(id, \*, bits)
+
+   Access the ADC peripheral identified by *id*, which may be an integer
+   or string.
+   
+   The *bits* argument, if given, sets the resolution in bits of the
+   conversion process.  If not specified then the previous or default
+   resolution is used.
+
+Methods
+-------
+
+.. method:: ADCBlock.init(bits)
+
+   Configure the ADC peripheral.  *bits* will set the resolution of the
+   conversion process.
+
+.. method:: ADCBlock.connect(channel)
+.. method:: ADCBlock.connect(source)
+.. method:: ADCBlock.connect(channel, source)
+
+   Connect up a channel on the ADC peripheral so it is ready for sampling,
+   and return a :ref:`machine.ADC <machine.ADC>` object that represents that
```

Comment:
Please remember this pattern for the next time I2C/SPI API is discussed. Because with I2C, it should be the same - there should be a way to create an "I2C connection" object, over which normal stream interface works.

Reply:
I2C doesn't work with a stream interface because it's inherently a message based bus.

Reply:
> I2C doesn't work with a stream interface because it's inherently a message based bus.

There's nothing in stream interface per se which precludes it working with a stream of messages. It's a simple, powerful, reusable, composable abstraction. (Though yes, currently, mostly stream of bytes mode is well exercised with it. And yes, there're potential drawbacks with too much generalization and packing too much semantics in one interface, but issues with inconsistent design is far more numerous and obvious.)